### ISSUE TITLE
Weird poor performance when running on IIS pool with 32bit apps enabled

### ISSUE THREAD

**Issue by ricardopieper** [StackExchange/StackExchange.Redis#437]
Hello!

**Tl;DR;** When enabling 32 bit applications on the IIS pool, StackExchange.Redis kind of slows down in a big way. Like, from 0.01 millseconds to 50~100 milliseconds. I provide more details in the non-TL;DR text, so yeah, this TL;DR is kinda useless.

First of all, I must say that SE.Redis is very fast. Doesn't matter how much I try to stress it, SE.Redis always wins. My tests are all done in a simple C# console app, where I managed to get 100k/ops with a 1.7ghz core i3 cpu against a redis server installed in the same windows machine, which is quite nice. We measure all the timings using the SE.Redis profiler interface, and we see everything happening in less than a millisecond.

I created a brand new ASP.NET application on Visual Studio to simulate the problem we're having, and I just added an action that calls SE.Redis on a GET request. It looks like this:

```c#
    [Route("DoStuffRedis")]
    [AllowAnonymous]
    public string GetDoStuffRedis()
    {
        Stopwatch sw = Stopwatch.StartNew();
        //testing without our wrapper
        var db = Multiplexer.Redis.GetDatabase();

        var text = Guid.NewGuid().ToString();

        int ops = 50;
        while (ops-- > 0)
        {
            db.StringGet("NUC001_109_0");
            db.StringSet("text", text);
        }
        sw.Stop();
        return "did stuff in " + sw.Elapsed.TotalMilliseconds;
    }
```

So far we've been using another redis library (you guessed it, ServiceStack.Redis) but we want to move to SE.Redis because of reasons not important to this discussion (although SS.Redis is pretty good too). 

In this case (and in this report) we're just using SE.Redis, but we actually have a wrapper around it, because of reasons. It doesn't do much though, it just calls IDatabase methods.

We fired up IIS Express and gave it a shot. As expected, SE.Redis delivered sub-millisecond operations. So far, so good!

Then I tried to put that application to run in an IIS environment. For our surprise, SE.Redis is performing in a very, very slow fashion, taking almost 50 milliseconds per request, with a redis server installed on localhost. And yes, we're storing the `ConnectionMultiplexer` object The Right Way™.

So I went for a coffee break and it just happens that my colleague was there. I told him about this since he is using SE.Redis in a development environment, and surprisingly enough he does not report the same lag I'm having, which is weird... However, in another unrelated conversation, he said that he had just changed the pool configuration to "Enable 32 bit applications" due to memory consumption problems.

It just happens that all other applications in our servers are already running in pools with 32bit apps enabled! Is that the cause of the slowdown? 

I got back to my laptop and checked the IIS configurations... and there it is, I had configured my IIS pool the same way our production server is set up. 32 bit applications are enabled, and StackExchange.Redis is taking a long time to perform an operation. We enabled our little profiler and we got this output:

```json
[{
    "Command": "SET",
    "CommandCreated": "01/07/2016 16:12:13",
    "CreationToEnqueued": 0.0135,
    "EnqueuedToSending": 0.0045,
    "SentToResponse": 10.1572,
    "ResponseToCompletion": 0.0165,
    "ElapsedTime": 10.1917,
}, {
    "Command": "GET",
    "CommandCreated": "01/07/2016 16:12:13",
    "CreationToEnqueued": 0.0139,
    "EnqueuedToSending": 0.0052,
    "SentToResponse": 7.9659,
    "ResponseToCompletion": 0.0165,
    "ElapsedTime": 8.0015,
}]
```

What is more curious about that is the fact that the method itself seems to take a little bit longer than 10 or 8ms: it's actually 50~100ms. I measured the times by just adding a stopwatch to `ExecSyncImpl<T>()` and measured the times taken by each line. Turns out it takes 50~100ms for `if (Monitor.Wait(source, timeoutMilliseconds))` alone to finish. The same lag happens with the NuGet package. We also used Stackify Prefix to measure these times:

![Stackify Prefix in action](https://cloud.githubusercontent.com/assets/4030858/16529082/39d43c88-3f97-11e6-819b-e71e01c0bf43.png)

(In this image we're actually using our wrapper, but yeah, you can see that it spends the whole time doing the redis operation, our wrapper is just a lightweight proxy).

So I went back to the pool settings and changed the config back to False, disabling 32 bit apps. And lo and behold, performance is great again. Nothing changed in our wrapper. The method executes in less than a millisecond, as reported by Stackify Prefix. Here's the profiler output:

```json
[{
    "Command": "SET",
    "CommandCreated": "01/07/2016 17:03:27",
    "CreationToEnqueued": 0.0078,
    "EnqueuedToSending": 0.0055,
    "SentToResponse": 0.0196,
    "ResponseToCompletion": 0.0089,
    "ElapsedTime": 0.0418,
}, {
    "Command": "GET",
    "CommandCreated": "01/07/2016 17:03:27",
    "CreationToEnqueued": 0.007,
    "EnqueuedToSending": 0.0054,
    "SentToResponse": 0.0209,
    "ResponseToCompletion": 0.0085,
    "ElapsedTime": 0.0418,
}]
```

So... yeah. I'll try to do more tests here, but I'm here to ask if you guys know something about this. It's kinda weird to me. Is it normal that the "Enable 32bit applications" setting = true would cause that kind of lag? We don't think this is machine-specific, since this happens in multiple computers and servers. 

If this is a problem, I might even send you more information, I just don't know what else can I provide, so feel free to ask.

Thank you for your attention!

**Comment by NickCraver** [comment]
Closing out to cleanup, but happy to revisit if anyone is still having issues.

**Comment by ricardopieper** [comment]
I also meant to answer this... I thought I did actually.

I don't remember what was the solution. We saw that there was a new version right afterwards that changed some async or I/O stuff I'm not sure. Just updating the lib got rid of that behavior.

I don't work at that company anymore so I don't know what's the status of this issue, but they had moved to 64bit anyway IIRC
